tests/php/ORM/ArrayListTest.php

```diff
@@ -1236,4 +1236,15 @@ public function testDataClass()
         $list->setDataClass(DataObject::class);
         $this->assertEquals(DataObject::class, $list->dataClass());
     }
+
+    public function testShuffle()
+    {
+        $list = new ArrayList(range(1, 10));
+
+        $list->shuffle();
+
+        for ($i = 1; $i <= 10; $i++) {
+            $this->assertContains($i, $list);
+        }
```

Comment:
The odds of this coming back in the same order are 1 / 3,600,000. Should we have it test the expected result (a shuffle) and allow 1 in 3.6M tests to fail curiously? If not, set the range to 50 and if the probability gods rain down up on us, we deserve it.

Reply:
but there's no criterion saying it must not repeat the original list. that can be considered as just 1 more possible combination

Reply:
Yeah I didn't bother testing that it actually shuffles - I feel like we're just testing PHP functionality at that point. If there was some way I could assert that the `shuffle` line is being called though... 

I'm happy to up the range (a lot) and add a `$this->assertNotEquals()` or whatever I need to do for array comparison 😧

Reply:
simply saying that ideally tests should reflect the functionality requirements "as is".
Otherwise, if we add such test, we may need to add that feature to the function implementation (e.g. calculate a hash of the list and after shuffling check the hash of the new list is different, otherwise repeat the shuffle).

Reply:
It's just weird that the test doesn't test what it's supposed to do. If I went in and changed the `shuffle()` method to `return $this;`, the tests would pass.

Reply:
agree. what if we test that 2 shuffles are not the same? :) I guess that's still out of our control (as RNG is hidden under php), but at least we don't exclude the `same as the original` variant from the possible outcomes?

Reply:
If we changed the range to 50, we would be approaching UUID entropies. A false negative is never, ever, ever, ever going to happen.

Reply:
I'm not that worried about false-positives, but rather thinking that tests should describe functional requirements. Currently, if we add a simple check that shuffled != original, then it won't be reflecting the truth, as we don't have this functionality implemented.

I'm only saying that, because I already had the problem in understanding whether test actually reflects requirements, or simply trying to check some side-effects. In particular this one - https://github.com/silverstripe/silverstripe-framework/blob/4/tests/php/ORM/DataListTest.php#L614

After a while, people maintaining the code base won't know whether we actually have this functional requirement, or do we just test it for having test coverage.

Reply:
```php
$shuffle1 = $list->shuffle();
$shuffle2 = $list->shuffle();

$shuffle1 !== $shuffle2
```

Does that not accurately describe a shuffle, barring one extremely, out-of-this-universe edge case?

Reply:
haha, sorry for being a pain... but yes, this appears to be what I imagine :)

Reply:
Ok I've updated the test. Asserting two shuffles are different isn't really any different from asserting a shuffle isn't the same as the original order - I've also up-ed the range to 50.